api/admin/routes.py

```diff
@@ -81,6 +80,21 @@ def decorated(*args, **kwargs):
         return f(*args, **kwargs)
     return decorated
 
+def has_library(f):
```

Comment:
I tried to reuse `api/routes.py`'s `has_library` decorator but it doesn't seem to work. I am getting an `AttributeError: 'Flask' object has no attribute 'manager'` error and it seems to be because on line 91, `app` is coming from `api/app` but in `api/routes.py` it is coming from the base. I'm getting stuck here because `app` is not being mocked as I expect it to, and as how the tests for the other api routes do mock `app` correctly.

Reply:
I think this is what's happening:

1. api/routes.py defines `app`.
2. api/admin/routes.py imports `app`, copying the object into its namespace.
3. Your test setup mocks `api/admin/routes.py:app` (i.e. changes which object it refers to) but leaves  `api/routes.py:app` alone.
4. Code in `api/admin/routes.py` calls decorator code in `api/routes.py` which references the un-mocked `api/routes.py:app`, causing the crash.

You should be able to fix this by mocking `api.routes.app` at the same time you mock `api.admin.routes.app`. They start the test as two references to the same real object, they should both be mocked (so they are references to the same mock object) and then they should both be unmocked (so they end up references to the same real object).

Reply:
That was it. Thanks.